### ISSUE TITLE
Compatibility with Webpack 4 / Create React App 2

### ISSUE THREAD

**Issue by wojtekmaj** [wojtekmaj/react-pdf#179]
Due to PDF.js being tightly coupled with worker-loader, at the moment we can't update the code so that it works with Webpack 4 flawlessly. It *might* work, but no guarantees can be made.

The issue has been filed against Mozilla's PDF.js in mozilla/pdf.js#9580 to separate Webpack specific code to another package, which would give us much more flexibility regarding fighting issues like that.

This issue is to be monitored.

**Comment by SavePointSam** [comment]
For users of Create React App 2, or anyone that is using webpack 4 in general. This issue appears as the following message when using the webpack entry point:

```
.../node_modules/pdfjs-dist/build/pdf.worker.js
Module build failed: TypeError: Cannot read property 'thisCompilation' of undefined
```

**Comment by SavePointSam** [comment]
@wojtekmaj I requested that a new release that is webpack 4 compatible be pushed out last week (https://github.com/mozilla/pdf.js/issues/9440#issuecomment-394847900). It appears that they did cut a release that may work (2.0.550). How can we test this version works on our end?

**Comment by wojtekmaj** [comment]
Hmmm, that's fairly easy. You can run project in `/sample` directory after upgrading Webpack to 4.x and doing minor changes to Webpack config (e.g. adding `mode`).

**Comment by wojtekmaj** [comment]
Another issue to monitor before it gets resolved: mozilla/pdf.js#9733

**Comment by SavePointSam** [comment]
Thanks! I'm going to take some time today to look into updating to the newer version as we discussed a few weeks ago.

**Comment by wojtekmaj** [comment]
Upgrading should be as easy as changing a dependency for a new PDF.js version which uses new worker-loader and consequently supports Webpack 4, but drops support for Webpack 2. We're just waiting for a official pdf.js 2.0 release.

**Comment by banyan** [comment]
@wojtekmaj I've just followed [your instruction](https://github.com/wojtekmaj/react-pdf/issues/179#issuecomment-397216778) to test webpack4. Forked [pdfjs-dist](https://github.com/banyan/pdfjs-dist) and [react-pdf](https://github.com/banyan/react-pdf/tree/pdfjs2) and it seems working well! (thanks for your work!!)
I'd like to clarify that it doesn't work if I specify `import { Document } from 'react-pdf/dist/entry.webpack';` instead of `import { Document } from 'react-pdf';`. does [this](https://github.com/wojtekmaj/react-pdf/wiki/Upgrade-guide-from-version-3.x-to-4.x#non-worker-version-is-gone) mean we don't need to import from `react-pdf/dist/entry.webpack` from version 4?

**Comment by wojtekmaj** [comment]
No, that does not mean we shouldn't use bundler-specific entry points. They are there to kickstart you real fast. Although if you have pdf.worker.js provided on your own in correct location, you can safely use the default entry.

I'm glad that Webpack 4 worked for you - I'll be upgrading it as soon as we get the confirmation from Mozilla that PDF.js 2.0 is ready to go.

**Comment by imperio59** [comment]
I ran into issues where the worker.js bundle was not being loaded. 

I fixed it by adding a publicPath entry in my output object in webpack config, as well as globalObject: 'this' as explained above. See https://webpack.js.org/guides/public-path/

**Comment by filgaia** [comment]
Any progress in this issue?

We got this error with Webpack 4 and react-pdf 3.X

![image](https://user-images.githubusercontent.com/30418424/46376897-db19eb80-c65c-11e8-8ec3-407a204ba245.png)

With version 4.X is not even loading our app. :(

**Comment by SavePointSam** [comment]
How is `webpack` being listed as in `peerDependencies` blocking progress on this?

**Comment by wojtekmaj** [comment]
Webpack not so much, but `worker-loader` is the key. In version 1.x, it's not compatible with Webpack 4, so you can't use Webpack-specific entry.

In PDF.js 2.x, which is used internally by React-PDF 4.x, this dependency was upgraded to 2.0, which should solve some issues with Webpack 4. You can test React-PDF 4.x now and feedback is very welcome.

**Comment by SavePointSam** [comment]
So in theory, React-PDF@^4.0.0 is react-scripts@^2.0.0 compatible?

**Comment by wojtekmaj** [comment]
@SavePointSam 

FYI all, we got info from Mozilla and React-PDF 4.0.0 final release should be out next week.

**Comment by barrymichaeldoyle** [comment]
Is that final release out yet?

**Comment by wojtekmaj** [comment]
No it's not. We're blocked by mozilla/pdf.js#10205. I'm considering releasing 4.0.0 final with pdfjs-dist 2.0.550 at this point.

**Comment by bw7432** [comment]
Any update on this? I'm getting the same error (Uncaught ReferenceError: window is not defined) in a react-electron-boilerplate project.

**Comment by varunpatro** [comment]
@wojtekmaj I've been trying to diagnose this for some time now, and I'm not sure how to proceed. 

> I'm considering releasing 4.0.0 final with pdfjs-dist 2.0.550 at this point.

Is this still going to happen?

What's the plan for now?

**Comment by wojtekmaj** [comment]
@varunpatro 4.0.0 final was released a long time ago now. It should work fine with Webpack 4. If it isn't, it should be considered a bug.

**Comment by brydenfogelman** [comment]
@wojtekmaj Does following warning go away with the most recent version?

`./node_modules/pdfjs-dist/build/pdf.js
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted`

**Comment by wojtekmaj** [comment]
@brydenfogelman No, this is tracked under #280 and is not closed.

**Comment by Maskik** [comment]
Looks like it resolved in [this solution](https://github.com/plangrid/react-file-viewer/blob/master/src/components/drivers/pdf-viewer.jsx)

try this:
```
import { PDFJS } from 'pdfjs-dist/build/pdf.combined';
import 'pdfjs-dist/web/compatibility';
```
_maybe you need to use lower version "pdfjs-dist": "1.8.357"_
